File: examples/client/tutorials/transactions/divisible_transactions.py
from bigchaindb_driver import (
    BigchainDB,
)
from bigchaindb_driver.crypto import generate_keypair

from cryptoconditions import Ed25519Fulfillment
from cryptoconditions.crypto import Ed25519VerifyingKey

from client.tutorials.constants.application_constants import BDB_SERVER_URL
from client.tutorials.utils.bigchaindb_utils import (
    poll_status_and_fetch_transaction,
    prepare_transfer,
    sign_ed25519
)

bdb = BigchainDB(BDB_SERVER_URL)

print(bdb.info())

alice = generate_keypair()
bob = generate_keypair()
carly = generate_keypair()

tx_create_alice_divisible = bdb.transactions.prepare(
    operation='CREATE',
    signers=alice.public_key,
    recipients=[([alice.public_key], 4)],
)

tx_create_alice_divisible_signed = bdb.transactions.fulfill(
    tx_create_alice_divisible, private_keys=alice.private_key)

asset_id = tx_create_alice_divisible_signed['id']

print('Posting signed transaction{}'.format(tx_create_alice_divisible_signed))
res = bdb.transactions.send(tx_create_alice_divisible_signed)
poll_status_and_fetch_transaction(tx_create_alice_divisible_signed['id'], driver=bdb)


tx_transfer_divisible = prepare_transfer(
        inputs=[
            {
                'tx': tx_create_alice_divisible_signed,
                'output': 0
            }
        ],
        outputs=[
            {
                'condition': Ed25519Fulfillment(public_key=Ed25519VerifyingKey(carly.public_key)),
                'public_keys': [carly.public_key],
                'amount': 2
            },
            {
                'condition': Ed25519Fulfillment(public_key=Ed25519VerifyingKey(bob.public_key)),
                'public_keys': [bob.public_key],
                'amount': 1
            },
            {
                'condition': Ed25519Fulfillment(public_key=Ed25519VerifyingKey(alice.public_key)),
                'public_keys': [alice.public_key],
                'amount': 1
            }
        ]
)

tx_transfer_divisible_signed = sign_ed25519(tx_transfer_divisible, alice.private_key)

print('Posting signed transaction{}'.format(tx_transfer_divisible_signed))
bdb.transactions.send(tx_transfer_divisible_signed)
poll_status_and_fetch_transaction(tx_transfer_divisible_signed['id'], driver=bdb)

res = bdb.transactions.get(asset_id=asset_id)
print('Retrieve list of transactions with asset_id {}: {}'.format(asset_id, len(res)))
# res = bdb.outputs.get(public_key=alice.public_key, unspent=True)
# print('Retrieve list of unspent outputs with public_key {}: {}'.format(alice.public_key, len(res)))
# res = bdb.outputs.get(public_key=bob.public_key)
# print('Retrieve list of unspent outputs with public_key {}: {}'.format(bob.public_key, len(res)))
# res = bdb.outputs.get(public_key=carly.public_key, unspent=True)
# print('Retrieve list of unspent outputs with public_key {}: {}'.format(carly.public_key, len(res)))

tx_transfer_divisible = prepare_transfer(
        inputs=[
            {
                'tx': tx_transfer_divisible_signed,
                'output': 0
            },
            {
                'tx': tx_transfer_divisible_signed,
                'output': 1
            }
        ],
        outputs=[
            {
                'condition': Ed25519Fulfillment(public_key=Ed25519VerifyingKey(bob.public_key)),
                'public_keys': [carly.public_key],
                'amount': 1
            },
            {
                'condition': Ed25519Fulfillment(public_key=Ed25519VerifyingKey(carly.public_key)),
                'public_keys': [bob.public_key],
                'amount': 2
            }
        ]
)

tx_transfer_divisible_signed = sign_ed25519(tx_transfer_divisible, [bob.private_key, carly.private_key])
print('Posting signed transaction{}'.format(tx_transfer_divisible_signed))
bdb.transactions.send(tx_transfer_divisible_signed)
poll_status_and_fetch_transaction(tx_transfer_divisible_signed['id'], driver=bdb)

res = bdb.transactions.get(asset_id=asset_id)
print('Retrieve list of transactions with asset_id {}: {}'.format(asset_id, len(res)))
# res = bdb.outputs.get(public_key=alice.public_key, unspent=True)
# print('Retrieve list of unspent outputs with public_key {}: {}'.format(alice.public_key, len(res)))
# res = bdb.outputs.get(public_key=bob.public_key)
# print('Retrieve list of unspent outputs with public_key {}: {}'.format(bob.public_key, len(res)))
# res = bdb.outputs.get(public_key=carly.public_key, unspent=True)
# print('Retrieve list of unspent outputs with public_key {}: {}'.format(carly.public_key, len(res)))

File: examples/client/tutorials/transactions/divisible_transactions.js
// Install necessary polyfills (see supported browsers) into global
import 'core-js/es6';
import 'core-js/stage/4';
import 'isomorphic-fetch';

import {
    Ed25519Keypair,
    makeCreateTransaction,
    makeTransferTransaction,
    makeOutput,
    makeEd25519Condition,
    signTransaction,
} from 'js-bigchaindb-quickstart';

import {
    postTransaction,
    pollStatusAndFetchTransaction,
    listTransactions,
    listOutputs
} from '../utils/bigchaindb_utils';


const alice = new Ed25519Keypair();
const bob = new Ed25519Keypair();
const carly = new Ed25519Keypair();

const txCreateAliceDivisible = makeCreateTransaction(
    {assetMessage: 'I will stick to every future transfer transaction'},
    {metaDataMessage: 'I am specific to this create transaction'},
    [makeOutput(makeEd25519Condition(alice.publicKey), 4)],
    alice.publicKey
);

// sign, post and poll status
const txCreateAliceDivisibleSigned = signTransaction(txCreateAliceDivisible, alice.privateKey);
const assetId = txCreateAliceDivisible.id;

console.log('Posting signed transaction: ', txCreateAliceDivisibleSigned);
postTransaction(txCreateAliceDivisibleSigned)
    .then((res) => {
        console.log('Response from BDB server', res);
        return pollStatusAndFetchTransaction(txCreateAliceDivisibleSigned)
    })
    .then((res) => {
        listTransactions({asset_id: assetId})
            .then((res) => {
                console.log('Retrieve list of transactions with asset_id', assetId, res);
            });

        listTransactions({asset_id: assetId, operation: 'create'})
            .then((res) => {
                console.log('Retrieve list of create transactions with asset_id', assetId, res);
            });

        listTransactions({asset_id: assetId, operation: 'transfer'})
            .then((res) => {
                console.log('Retrieve list of transfer transactions with asset_id', assetId, res);
            });
    });
